lecture-notes/l1-intro-new.md

```
@@ -0,0 +1,439 @@
+
+# 3 From Python to Java
+ 
+Both of the above implementations in Python have some limitations:
+
+  1. In the first implementation, we need to know what kind of a device has been passed to the `identify()` function so that it can work accordingly. Doing so is rather clunky: our helper functions basically differentiate between devices based on the contents of the dictionary representation. *Any* dictionary that happens to have the correct attributes will be mistaken as a device, whether or not it is meaningfully so.
+
+  2. In both implementations, there is no *succinct* way to know what the capabilities of devices are. The `IoTDevice` class in the second implementation does show the functions available to all devices, but it mixes them with the common attributes. 
+
+  3. Python allows reaching into an object to access its attributes. This encourages (or at least allows) writing code that is heavily intertwined. This makes maintaining code over a period of time more challenging.
+
+## 3.1 Replicating the object-oriented design in Java
```

markdownlint-cli2 (0.22.0)

[warning] Heading style
Expected: setext; Actual: atx

(MD003, heading-style)